### PR TITLE
Add support to verifying the signature

### PR DESCRIPTION
**Pull request by mpistrang**
Here are the change I am proposing for the apple pay signature verification.  This is a PR from my verify_signature branch to mpistrang:master since the code is not final and I'm looking for feedback on the following before submitting a PR for Cashstar's fork.

Feedback items:
- the skipped tests cover functions that I'm not sure how best to test yet, specifically `get_signed_data`
- should the top level verify_signature be on the Payment object that we used to decode?
- Any/all other suggestions on code organization or tests, there is a lot here and I'd like to make it as clean as possible.

### CODE REVIEW

File: applepay/utils.py [hunk 1]
```diff
@@ -196,70 +202,90 @@ def get_provided_data(token):
...
+            algorithm updated with the data to be hashed
     """
     hashfunc = hashlib.new(name)
     hashfunc.update(data)
     return hashfunc
 
 
-def signature_is_valid(token):
-    """TODO write me
+def verify_signature(token):
```
Comment (jmsmalley) [inline applepay/utils.py]: I think we should be using function `signing_time_is_valid` from line 56 somewhere inside of `verify_signature` to meet the requirement from apple docs 
`Inspect the CMS signing time of the signature, as defined by section 11.3 of RFC 5652. If the time signature and the transaction time differ by more than a few minutes, it's possible that the token is a replay attack.`

File: applepay/utils.py [hunk 2]
```diff
@@ -66,3 +85,295 @@ def signing_time_is_valid(signature, current_time, threshold):
...
+        return False
+
+    root_der = open(payment.ROOT_CA_FILE, 'r').read()
+
+    # pass the der-encoded representation of the certs
+    if not valid_chain_of_trust(root_der, intermediate_cert.dump(), leaf_cert.dump()):
+        return False
+
+    # Build the signer information from the signer that matches the leaf cert
+    signer = get_first_from_iterable(
```
Comment (jmsmalley) [inline applepay/utils.py]: The word `signer` sounds weird to me as I never heard of it. Is this short for signature?

File: applepay/utils.py [hunk 3]
```diff
@@ -66,3 +85,295 @@ def signing_time_is_valid(signature, current_time, threshold):
...
+    # build the hashfunc from the leaf_cert's defined algorithm
+    hashfunc = partial(get_hashfunc_by_name, leaf_cert.hash_algo)
+
+    hashed_data = hashfunc(data).digest()
+
+    if hashed_data != message_digest:
+        logger.warning("Message digest does not match provided data.")
+        return False
+
+    signed_data = get_signed_data(signed_attrs)
```
Comment (jmsmalley) [inline applepay/utils.py]: We are already using `signed_data` variable above. Seems weird to re-use it.

File: applepay/utils.py (hunk 2 above)
Comment (mpistrang) [reply applepay/utils.py]: Maybe signee is a better word here.  What do you think?

File: applepay/utils.py (hunk 3 above)
Comment (mpistrang) [reply applepay/utils.py]: Good point, I will update this to not collide.

File: applepay/utils.py (hunk 1 above)
Comment (mpistrang) [reply applepay/utils.py]: I was thinking that too.  I'll think about how best to include it.

File: applepay/utils.py (hunk 2 above)
Comment (mpistrang) [reply applepay/utils.py]: Updated to `signer_info` since its one instance of of the SignerInfo class

File: applepay/utils.py (hunk 3 above)
Comment (mpistrang) [reply applepay/utils.py]: I update this function and variable name to be more closely tied to what its doing; returning the ber-encoded signed attributes.

File: applepay/utils.py (hunk 1 above)
Comment (mpistrang) [reply applepay/utils.py]: I just pushed a bunch of changes related to this, check it out.

**Comment by mpistrang** [comment]
I am going to merge this into my master branch and officially open the PR into cashstar:master

**Pull request merged**